File: MorrisInorderTraversal.java
public class MorrisInorderTraversal {
    private class TreeNode{

        int val;
        TreeNode left;
        TreeNode right;

        public TreeNode(){
             this.val=0;
             this.left=null;
             this.right=null;
        }

        public TreeNode(int val){
            this.val=val;
            this.left=null;
            this.right=null;
        }
    }

    public void inOrderTraversal(TreeNode root){
        TreeNode curr=root;
        while(curr!=null){
            if(curr.left==null){
                System.out.println(curr.val);
                curr=curr.right;
            }else{
                TreeNode prev=curr.left;
                while(prev.right!=null && prev.right!=curr){
                    prev=prev.right;
                }
                if(prev.right==null){
                    prev.right=curr;
                    curr=curr.left;
                }else{
                    System.out.println(curr.val);
                    prev.right=null;
                    curr=curr.right;
                }
            }
        }
    }
}

File: morris inorder traversal.py
#TC: O(n) in average case as we TC is constant for all the leaf nodes and so on...
#SC: O(1)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def isValidBST(self, root):

        def inorder(root):
            curr=root
            while(curr!=None):
                if curr.left==None:
                    print(curr.val)
                    curr=curr.right
                else:
                    pre=curr.left
                    while(pre.right!=None and pre.right!=curr):
                        pre=pre.right
                    if pre.right==None:
                        pre.right=curr
                        curr=curr.left
                    else:
                        print(curr.val)
                        pre.right=None
                        curr=curr.right
                        
        inorder(root)
        return True
        
                
        
        
        
            
        
        